Refactored the code a bit to be more readable, gearing up for mutithreading

fheinz opened the pull request:
Mostly reformatting of the code, and modularising it so it will be easier to split for multi-threading.

fheinz committed:
Update main.ino

Refactored the loop into smaller bits. Still needs work, and to rely a lot less on Sleep() calls for timing.

Also fixed Serial initialization, which was missing the baud rate.

fheinz committed:
Fix immediate shutdown and lack of serial output

* Fixed Serial initialization.
* The battery low indication counter check was inverted.

fheinz committed:
Pump up the power!

Set the power to the motors to almost full. This is something that will ideally be configurable in the future.

fheinz committed:
Readability improvements

Group related functions , introduce a few constants, that kind of thing.

Makes for a large change that doesn't do much in terms of code, unfortunately.

matthewlai commented on firmware/main/main.ino:
```diff
@@ -231,107 +231,104 @@ void Shutdown(bool with_flashing_leds = true) {
+      Serial.println("Found a device! Validating...");
+      client->setClientCallbacks(new MyClientCallback());
+      client->connect(&remote_device);
+      BLERemoteService* remote_service = client->getService(SERVICE_UUID); //do we ever need to free this?
```
I believe the lifetime of these objects are all managed by the client, so we don't need to free them.
https://h2zero.github.io/NimBLE-Arduino/class_nim_b_l_e_client.html#a6b801e91a37745445499dadfb4c38597

matthewlai replied:
"new MyClientCallback" above does leak memory though. We should probably just reuse the instance.

matthewlai commented on firmware/main/main.ino:
```diff
@@ -231,107 +231,104 @@ void Shutdown(bool with_flashing_leds = true) {
+    }
+  }
+  scan->clearResults();
+  delay(2000);
```
We don't need to wait if we are now connected. This is just to insert some dead time between scans to save some battery. I guess this is part of why the connection is on startup is so slow (this accounts for 2 seconds!)

matthewlai commented on firmware/main/main.ino:
```diff
@@ -237,14 +237,14 @@ void Shutdown(bool with_flashing_leds = true) {
   esp_deep_sleep_start();
 }
 
+float vbatt = ReadVBatt();
```
This needs to happen every iteration

matthewlai commented on firmware/main/main.ino:
```diff
@@ -264,7 +278,7 @@ void ServerLoop() {
 
-bool EnsureConnectionToServer() {
+bool EnsureBLEConnectionToServer() {
   if (last_device_connected) return true;
```
Should be connected_to_server I think. last_device_connected is for the server.

matthewlai commented:
Thanks!

fheinz committed:
Multi-threading implementation

Separate BLE and motor control into separate threads.

This seems to work more reliably. I don't think we're quite where we want in terms of synchronization, need to work out a way to make sure both gloves are on the same page about the pattern in a way that is more by design than by chance.
